### ISSUE TITLE
Missing information in merged AnnData objects

### ISSUE THREAD

**Issue by sjspielman** [AlexsLemonade/scpca-nf#681]
While working towards https://github.com/AlexsLemonade/scpca-docs/issues/241, I have identified some discrepancies in the exported `AnnData` objects. We might want to split this issue up, but getting all my thoughts down here to start.

###  Discrepancy 1
The `.obs` slot is missing two expected columns: `assay_ontology_term_id` and `suspension_type`. However, the column `seq_unit` does exist in the object; it just didn't make it into the `suspension_type` column too. 

Here is where we add this information to `AnnData`:
https://github.com/AlexsLemonade/scpca-nf/blob/c0e257b2d6c5eb438e1677737393f5d4ca28e258/bin/sce_to_anndata.R#L82-L91

But, this won't work for merged objects, because we have placed all these metadata columns into library-specific data frames stored as `metadata(merged_sce)$library_metadata$SCPCL000000`. As a consequence, this information is missing from `.obs`. 

#### Solution to discrepancy 1

For _merged objects_, we already have added this information to `colData` already before getting to `AnnData` conversion:
https://github.com/AlexsLemonade/scpca-nf/blob/c0e257b2d6c5eb438e1677737393f5d4ca28e258/bin/merge_sces.R#L278-L287

Therefore, during `AnnData` conversion, we don't actually need to bring `assay_ontology_term_id` or `seq_unit` into the `colData`. We just need to add `is_primary_data = FALSE`. 

I suggest we add a check for if we're working with a merged object, and if so, only run: `sce$is_primary_data = FALSE` without messing with any other `colData` columns. We could add a check in a few ways, e.g.:
- a flag to the script itself
- check if, for example, `metadata(sce)$library_metadata`  is not `NULL`

###  Discrepancy 2

It turns out we actually have no `library_metadata` represented anywhere in the `AnnData` object! This information is totally lost in the conversion, since we have to remove all list & df objects from the sce metadata in `scpcaTools`:
https://github.com/AlexsLemonade/scpcaTools/blob/cec26960ed3fbfc852146e801f7c82c42c50d1b2/R/sce_to_anndata.R#L64-L67

I do think it's worth finding a way to retain this information, which you can see laid out in this table: https://github.com/AlexsLemonade/scpca-docs/blob/3afe4ae3fb4ab356d96caa8cf5298a7058065bb2/docs/merged_objects.md?plain=1#L127-L174

#### Solution to discrepancy 2

What if we just port these columns into the `colData`? Again, we'd want to leverage a check in `sce_to_anndata.R` for whether we're working with a merged object. Other than stashing these in `colData`, I don't have any particular ideas that will appease an SCE -> AnnData conversion.

Tagging @allyhawkins @jashapiro for discussion/thoughts

**Comment by jashapiro** [comment]
For the first one, do we need `is_primary_data` at all? Can we skip it since I don't think we care about CZI compliance for merged objects? 

For the second, I suppose adding columns makes sense, but not all items will make it in, as we would only be able to add those that are single values (there are a few data frames and objects in the metadata there)

**Comment by sjspielman** [comment]
> Can we skip it since I don't think we care about CZI compliance for merged objects?

If we don't care about CZI compliance, definitely don't need it at all. 

> For the second, I suppose adding columns makes sense, but not all items will make it in, as we would only be able to add those that are single values (there are a few data frames and objects in the metadata there)

This makes sense to me.


Any particular preferences for an "is this a merged object" check? I sort of prefer a script flag, but I don't feel strongly if you have a stronger opinion.

**Comment by jashapiro** [comment]
A script flag seems fine to me.

**Comment by sjspielman** [comment]
> > Can we skip it since I don't think we care about CZI compliance for merged objects?
>
> If we don't care about CZI compliance, definitely don't need it at all.

Actually a broader question here then, is whether we even want to pass merged objects into the `format_czi()` function at all. We could instead have an `if` in the body of the script where merged objects get processed differently altogether, potentially in a separate function `format_merged_objects` or so (depending on how many lines of code it turns out to be). For example, we don't also need to add `schema_version`, `feature_is_filtered`, or append reducedDims with `X_`.

https://github.com/AlexsLemonade/scpca-nf/blob/c0e257b2d6c5eb438e1677737393f5d4ca28e258/bin/sce_to_anndata.R#L59-L113

I think this is probably my preference on where to head with this.

**Comment by allyhawkins** [comment]
Although we don't need to be CZI compliant here, I do think we should be as consistent as possible with how we store things in merged vs individual AnnData objects. 
We could get rid of things that we don't care about, `feature_is_filtered`, `schema`, and `is_primary_data`, but I think we want to keep naming consistent, so keep the `X_` prefix. 

I think maybe just having a separate function for `format_merged_anndata` and then if the object is merged run that function and if not run the other function. 

I agree about adding some, but definitely not all of the library metadata columns. I'm torn on if we actually need any of those columns though and can redirect users to the metadata file they download that will have all of that information.

**Comment by sjspielman** [comment]
> We could get rid of things that we don't care about, feature_is_filtered, schema, and is_primary_data, but I think we want to keep naming consistent, so keep the X_ prefix.

I agree with `X_`.

> I think maybe just having a separate function for format_merged_anndata and then if the object is merged run that function and if not run the other function.

This is my thought! I'm working on it now. After getting rid of all objects and dfs, these fields remain:
```
 [1] "library_id"              "sample_id"              
 [3] "project_id"              "salmon_version"         
 [5] "reference_index"         "total_reads"            
 [7] "mapped_reads"            "mapping_tool"           
 [9] "alevinfry_version"       "af_permit_type"         
[11] "af_resolution"           "usa_mode"               
[13] "af_num_cells"            "tech_version"           
[15] "assay_ontology_term_id"  "seq_unit"               
[17] "transcript_type"         "include_unspliced"      
[19] "filtering_method"        "prob_compromised_cutoff"
[21] "scpca_filter_method"     "min_gene_cutoff"        
[23] "normalization"           "highly_variable_genes" 
```

Some of these are already reflected in colData (e.g. `library_id`), so we can just remove them. Others are single strings, which seems ok to keep. Something like `highly_variable_genes` I would say keep, but collapse to a single either comma or semicolon separated string.

**Comment by sjspielman** [comment]
> Edit - woops, pressed enter too soon. But I think it's a good point that much of this is in the metadata, so maybe I'll compare and pare down to what they can't find elsewhere?

I've convinced myself that basically everything is the metadata. So, maybe just append that `X_` and call it a day?

**Comment by allyhawkins** [comment]
> I've convinced myself that basically everything is the metadata. So, maybe just append that `X_` and call it a day?

I am good with this.

**Comment by allyhawkins** [comment]
I just noticed this comment when reviewing https://github.com/AlexsLemonade/scpca-docs/pull/260

> Note that I have one !!TODO! little bit in there since whether we end up with suspension_type (vs just keeping it as seq_unit) depends on that https://github.com/AlexsLemonade/scpca-nf/issues/681 as well. I imagine we'll be keeping it as seq_unit since we don't need CZI compliance for merged objects?

I think we actually do want to keep the naming of this uniform so we should change to use `suspension_type` if that's what we are using in AnnData for individual objects. We should probably make this adjustment when we add that column to the merged object and not in the conversion to AnnData script.

**Comment by sjspielman** [comment]
>  if that's what we are using in AnnData for individual objects. We should probably make this adjustment when we add that column to the merged object and not in the conversion to AnnData script.

I'll update the column name, but FYI column _is_ added in the conversion to anndata script - https://github.com/AlexsLemonade/scpca-nf/blob/cd4499b1b3cacf5bb48ae32f005c05b27d84dc97/bin/sce_to_anndata.R#L75-L84
The column name in SCEs is still `seq_unit`.

**Comment by allyhawkins** [comment]
I'm saying that we should change what's in the SCE to be consistent so we aren't switching the naming between SCE and AnnData. This column only exists in merged SCE's and we should use `suspension_type`. So change this line: 
https://github.com/AlexsLemonade/scpca-nf/blob/c0e257b2d6c5eb438e1677737393f5d4ca28e258/bin/merge_sces.R#L298

**Comment by sjspielman** [comment]
> I'm saying that we should change what's in the SCE to be consistent so we aren't switching the naming between SCE and AnnData.

Ah, I see what you mean. So, since we still have this difference for single objects, in my mind it was "consistent" to keep `seq_unit` for SCE and `suspension_type` for AnnData. If we make them the same for merged, I would think we might want to make it the same for single objects too?

**Comment by allyhawkins** [comment]
> > I'm saying that we should change what's in the SCE to be consistent so we aren't switching the naming between SCE and AnnData.
> 
> Ah, I see what you mean. So, since we still have this difference for single objects, in my mind it was "consistent" to keep `seq_unit` for SCE and `suspension_type` for AnnData. If we make them the same for merged, I would think we might want to make it the same for single objects too?

But we don't include the `seq_unit` column for single objects? It's only in the metadata, but it is not a column in the `colData` in individual SCEs.

**Comment by sjspielman** [comment]
> But we don't include the seq_unit column for single objects? It's only in the metadata, but it is not a column in the colData in individual SCEs.

ugh there are so many slots.. Final.final destination:

- In single SCEs, we'll have `seq_unit` in metadata
- In merged SCEs, we'll leave the `seq_unit` that exists in each `metadata(sce)$library_metadata` as is. But change the `colData` version of that column to `suspension_type`
- In single and merged AnnDatas, we have `suspension_type` in `.obs`

**Comment by sjspielman** [comment]
Closed with #683
